pinot-core/src/main/java/com/linkedin/pinot/core/common/DataFetcher.java

```diff
@@ -43,11 +43,24 @@
   // Map from MV column name to max number of entries for the column.
   private final Map<String, Integer> _columnToMaxNumMultiValuesMap;
 
-  // Re-usable array for all dictionary ids in the block, of a single valued column
-  private final int[] _reusableDictIds;
+  // Thread local (reusable) array for all dictionary ids in the block, of a single valued column
+  private static final ThreadLocal<int[]> THREAD_LOCAL_DICT_IDS = new ThreadLocal<int[]>() {
+    @Override
+    protected int[] initialValue() {
+      return new int[DocIdSetPlanNode.MAX_DOC_PER_CALL];
+    }
+  };
 
   // Re-usable array to store MV dictionary id's for a given docId
-  private final int[] _reusableMVDictIds;
+  private static final ThreadLocal<int[]> THREAD_LOCAL_MV_DICT_IDS = new ThreadLocal<int[]>() {
```

Comment:
question is do we even need to lower it if the server is going to serve similar data and queries ? 
Frequency of expand/shrink is important otherwise whole change is not useful.
Two options:
1. If the expanded buffer will be one-off case then use that temporarily
2. Keep accounting information in get() call. Adjust based on values seen in last n accesses. But most of this will be counter-productive for GC goal